**Maximum Sub Sequence Problem**

Gegeben ist eine Liste welche die z.B. die Kursschwankungen einer Aktie fürht. Um den bestmöglichen Kauf- und Verkaufszeitpunkt zu berechnen, muss also das Maximum aus den Summen, die sich aus allen möglichen Kombinationen von  Start- und End-Indizes ergeben, berechnet werden.

In [1]:
import random
import time

In [2]:
def findmax(l):
    res = []
    for i in range(len(l)):
        for j in range(i,len(l)):
            res.append(l[j]-l[i])
    return(max(res))

Nachdem ich verschiedene Lösungsansätze versucht habe bin ich auf die Funktion oben gekommen. Diese sollte eigentlich mehr oder weniger der mathematischen Funktion auf dem Aufgabenblatt entsprechen. Ich gehe davon aus, dass durch die Verwendung von zwei for-Schleifen, das ganze bei grösseren Listen nicht mehr sehr performant sein wird.

In [3]:
# Beispiel aus Aufgabenblatt
findmax([31, -41, 59, 26, -53, 58, 97, -93, -23])

150

In [4]:
# Beispiel, bei dem ein Gewinn möglich ist, da Wert immer sinkt
# da wir auch den gewinn berechnen wenn verkauf und zum selben zeitpunkt verkauft wird (was immer 0 ist),
# ist dies somit der höchste gewinn
findmax([100, 91, 82 , 73, 64, 55, 46, 37, 28, 19])

0

In [5]:
findmax([12,24,35,46,57,68,79,984,2345,546456,3445345])

3445333

In [9]:
# beliebige beispiele gleicher länge
sample = random.sample(range(-100,100),20)
print(sample)
findmax(sample)

[-63, 13, -88, -42, 73, 12, 67, 28, -82, 62, 63, -2, 9, 65, 53, -33, -60, 31, -54, -39]


161

**Wie kann man diese Funktion testen?**

Verschiedene Listen (grösse der Liste  und Werte) auf die Funktion anwenden. Das Funktion sollte 
1. Kein Fehler ausgeben
2. Die richtige (höchste) Summe wird ausgegeben

**Welche Situationen sollte man testen?**

- Beliebige Liste
- Nur Werte welche immer mehr ins negative gehen
- Werte welche immer mehr in positive gehen
- Werte die immer gleich bleiben
- Liste mit nur 0
- Leere Liste


**Funktioniert die Funktion in allen Fällen korrekt?**

Soweit ja

In [68]:
# Laufzeit messen

def findmax_time(n,repeats):
    i = 0
    duration = 0
    while i < repeats:
        sample = random.sample(range(-1000000,1000000),n)

        start = time.time()
        res = findmax(sample)
        duration = duration + time.time() - start
        i = i + 1
    return duration/repeats
    


In [77]:
findmax_time(9,100)

7.3552131652832035e-06

In [79]:
findmax_time(20,100)

2.3107528686523438e-05

In [80]:
findmax_time(100,100)

0.00045601129531860354

In [81]:
findmax_time(1000,100)

0.04106937646865845

In [83]:
findmax_time(2000,100)

0.16994285106658935

In [84]:
findmax_time(4000,100)

0.6696550083160401

Die Laufzeit zu den Anzahl Elementen in der Liste ist nicht linear. 2000 Elemente benötigen 0.17 Sekunden, wobei die doppelte Listengrösse, also 4000 Elemente bereits 0.67 Sekunden benötigt, also etwa 4x so lange.